### PR TITLE
feat: make data types configurable and use FEEL per default

### LINKED ISSUE #674: Add FEEL data types
__What should we do?__

* [x] add the options time + dateTime + dayTimeDuration + yearMonthDuration + Any on selecting the Variable Type of inputs and outputs - in order to cover the most common types of FEEL
  * [x] for decision table
  * [x] for literal expression
* [x] replace several number types with `number`
* [x] reconsider simple edit dialog
  * [x] simple date edit
  * [x] simple time edit
  * [x] simple dateTime edit
  * [x] simple yearMonthDuration edit
  * [x] simple dayTimeDuration edit

__Why should we do it?__

We want to improve DMN support.

### LINKED ISSUE #677: Make data types configurable
* [ ] data types are configurable
* [ ] per default there are only FEEL data types

### PR DESCRIPTION
**Pull request by barmac**
Pass a list of strings via `config.dataTypes` to configure
available data types.

Camunda Platform-specific types have been replaced with FEEL equivalents.
Also, duration types (`dayTimeDuration` and `yearMonthDuration`)
as well as `Any` have been added.

Closes #674
Closes #677

BREAKING CHANGES:

* Types `integer`, `double`, and `long` have been replaced with `number`.
* Simple edit for `date` uses now FEEL date. For FEEL `date and time`,
  use `dateTime` type.

### CODE REVIEW

**Comment by barmac** [comment]
Before we merge this, let's discuss what we do regarding the simple edit. Already existing dialogs are not a 1:1 match with FEEL data types. Also, I am not super sure whether to keep the types as in XML or rather have readable labels:

image

image

**Comment by barmac** [comment]
Check out the build artifacts to test this easily in action:

- https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/677-feel-data-types/camunda-modeler-677-feel-data-types-linux-x64.tar.gz
- https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/677-feel-data-types/camunda-modeler-677-feel-data-types-mac.dmg
- https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/677-feel-data-types/camunda-modeler-677-feel-data-types-mac.zip
- https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/677-feel-data-types/camunda-modeler-677-feel-data-types-win-ia32.zip
- https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/677-feel-data-types/camunda-modeler-677-feel-data-types-win-x64.zip

**Comment by barmac** [comment]
I don't know what's wrong with the codecov, but it's related to https://github.com/bpmn-io/dmn-js/issues/672

**Comment by barmac** [comment]
Hmm it seems the simple edit additions are not available in the artifacts. I'll let you know when it's ready.

**Comment by barmac** [comment]
@nikku @pinussilvestrus @MaxTru Ready for review. The artifacts will be ready as soon as this job is finished: https://github.com/camunda/camunda-modeler/runs/5216236577?check_suite_focus=true

**Comment by MaxTru** [comment]
Not caused by this, but found during testing: https://github.com/bpmn-io/dmn-js/issues/685

**Review by MaxTru — CHANGES_REQUESTED** [review]
Some manual testing done:

* the simple edit validation for `yearMonthDuration`, `dayTimeDuration` is a bit too strict for `-PT5H` IMO (c.f. samples taken from https://camunda.github.io/feel-scala/docs/reference/builtin-functions/feel-built-in-functions-temporal/) (But I am not super comfortable with this FEEL data types, can you suggest some good docs?)
Screenshot_20220218_104316
* Same as above for date and time. Some examples from https://camunda.github.io/feel-scala/docs/reference/builtin-functions/feel-built-in-functions-temporal/ I cannot insert -> do I miss something?
Screenshot_20220218_105733

**Comment by barmac** [comment]
Thanks Max for looking into this.

For the time-zoned dates and times (`08:00:00@Europe/Berlin`), I think it makes sense to allow also other timezones than UTC.

However, I don't think we should allow negative durations (`-PT5H`). While they can be used in an example of absolute value function, they are nonsensical (nothing has ever lasted minus five hours).

**Comment by MaxTru** [comment]
Makes sense! Agreed.

What about `2015-09-18T10:31:10+01:00`?

**Comment by barmac** [comment]
Same. I will implement that next week.

**Comment by barmac** [comment]
OK so now we allow each of the following:
```
'08:00:00Z'
'08:00:00+01:00'
'08:00:00-01:00'
'08:00:00@Europe/Berlin'
'08:00:00'
```

**Comment by barmac** [comment]
TODO:
* change validation error color to red
* display invalid value if successfully unwrapped (e.g. `time("foo")`)
* keep first value on duration type change

**Comment by barmac** [comment]
@pinussilvestrus @MaxTru Ready for review.

File: packages/dmn-js-decision-table/src/features/type-ref/components/TypeRefCellContextMenu.js [hunk 1]
```diff
@@ -21,6 +12,7 @@ export default class TypeRefCellContextMenu extends Component {
 
     this._translate = context.injector.get('translate');
     this._modeling = context.injector.get('modeling');
+    this._dataTypes = context.injector.get('dataTypes');
```
Comment (MaxTru) [inline packages/dmn-js-decision-table/src/features/type-ref/components/TypeRefCellContextMenu.js:15]: Super cool

**Review by MaxTru — CHANGES_REQUESTED** [review]
I can't get the configuration of dataTypes to work. So what I try is the following:

```
new Modeler({
      container: container,
      common: {
        keyboard: {
          bindTo: document
        }
      },
      dataTypes: [
        'string',
        'boolean',
        'integer',
        'long',
        'double',
        'date',
        'mySuperCustomType'
      ]
    });
```

But I still end up with the default data types. I don't think we have an integration test for that either?

**Comment by barmac** [comment]
I think this has to be passed to `common` but will need to have a separate test indeed.

File: packages/dmn-js-shared/src/features/data-types/DataTypes.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  'Any'
+];
+
+/**
+ * Provide data types via config.
+ *
+ * @example
+ *
+ * // The data types will include multiple number types: integer, long, and double.
+ * const editor = new DmnJS({
```
Comment (MaxTru) [inline packages/dmn-js-shared/src/features/data-types/DataTypes.js:19]: See last comment, I suppose this documentation might be wrong

**Comment by barmac** [comment]
OK I have tested this and indeed the data types should be configured via `common`. I will update the JSDoc.

File: packages/dmn-js-shared/src/features/data-types/DataTypes.js [hunk 3]
```diff
@@ -17,14 +17,16 @@ const DEFAULT_DATA_TYPES = [
...
  * const editor = new DmnJS({
- *   dataTypes: [
- *     'string',
- *     'boolean',
- *     'integer',
- *     'long',
- *     'double',
- *     'date'
- *   ]
+ *   common: {
```
Comment (MaxTru) [inline packages/dmn-js-shared/src/features/data-types/DataTypes.js:20]: why not also have an integration test case for this (e.g., `should load customData types passed via <common.dataTypes>`)?

Comment (barmac) [reply packages/dmn-js-shared/src/features/data-types/DataTypes.js:20]: OK but such test will need to be added to `dmn-js` package. I will do that.

Comment (barmac) [reply packages/dmn-js-shared/src/features/data-types/DataTypes.js:20]: Done via https://github.com/bpmn-io/dmn-js/pull/678/commits/af2cb2c08840ec9eb26a6c6075817251e1675dab

Comment (MaxTru) [reply packages/dmn-js-shared/src/features/data-types/DataTypes.js:20]: :+1:

**Review by MaxTru — APPROVED** [review]
Great job!

**Pull request merged**